core/src/main/java/org/apache/calcite/rel/rules/SortRemoveDuplicateKeysRule.java

```diff
@@ -0,0 +1,134 @@
+
+/**
+ * Planner rule that remove duplicate sort keys.
+ *
+ * <p>The original SQL:
+ * <pre>{@code
+ * SELECT d1 FROM (
+ *   SELECT deptno AS d1, deptno AS d2 FROM dept
+ * ) AS tmp ORDER BY d1, d2
+ * }</pre>
+ *
+ * <p>The original logical plan:
+ * <pre>
+ * LogicalProject(D1=[$0])
+ *   LogicalSort(sort0=[$0], sort1=[$1], dir0=[ASC], dir1=[ASC])
+ *     LogicalProject(D1=[$0], D2=[$0])
+ *       LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ * </pre>
+ *
+ * <p>After optimization:
+ * <pre>
+ * LogicalProject(D1=[$0])
+ *   LogicalProject(DEPTNO=[$0], DEPTNO0=[$0])
+ *     LogicalSort(sort0=[$0], dir0=[ASC])
+ *       LogicalProject(DEPTNO=[$0])
+ *         LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ * </pre>
+ */
+@Value.Enclosing
+public class SortRemoveDuplicateKeysRule
+    extends RelRule<SortRemoveDuplicateKeysRule.Config>
+    implements TransformationRule {
+
+  /** Creates a SortRemoveDuplicateKeysRule. */
+  protected SortRemoveDuplicateKeysRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Sort sort = call.rel(0);
+    Project project = call.rel(1);
+    RelBuilder relBuilder = call.builder();
+
+    Map<RexNode, List<Integer>> exprToIndices = new LinkedHashMap<>();
+    List<RexNode> projects = project.getProjects();
+    for (int i = 0; i < projects.size(); i++) {
+      exprToIndices.computeIfAbsent(projects.get(i), k -> new ArrayList<>()).add(i);
+    }
+
+    List<RexNode> distinctProjects = new ArrayList<>(exprToIndices.keySet());
+    if (distinctProjects.equals(projects)) {
+      // No duplicate keys found, no need to transform.
+      return;
+    }
+
+    Map<Integer, Integer> oldToNewIdx = new HashMap<>();
+    int newIdx = 0;
+    for (List<Integer> idxList : exprToIndices.values()) {
+      for (int oldIdx : idxList) {
+        oldToNewIdx.put(oldIdx, newIdx);
+      }
+      newIdx++;
+    }
+
+    List<RelFieldCollation> newCollations =
+        sort.getCollation().getFieldCollations().stream()
+            .map(fc -> fc.withFieldIndex(oldToNewIdx.get(fc.getFieldIndex())))
+            .collect(Collectors.toList());
+
+    List<RexNode> newProjects = new ArrayList<>();
+    RexBuilder rexBuilder = sort.getCluster().getRexBuilder();
+    for (int i = 0; i < sort.getRowType().getFieldCount(); i++) {
+      RelDataTypeField field = sort.getRowType().getFieldList().get(i);
+      newProjects.add(rexBuilder.makeInputRef(field.getType(), oldToNewIdx.get(i)));
+    }
+
+    relBuilder.push(project.getInput())
+        .project(distinctProjects)
+        .sort(RelCollations.of(newCollations))
+        .project(newProjects);
+
+    call.transformTo(relBuilder.build());
+  }
+
+  /** Rule configuration. */
+  @Value.Immutable
+  public interface Config extends RelRule.Config {
+    Config DEFAULT = ImmutableSortRemoveDuplicateKeysRule.Config.of()
+        .withOperandSupplier(b0 ->
+            b0.operand(Sort.class).oneInput(b1 ->
```

Comment:
what if the input is not a project?
Can the sort specify the same column twice without using an alias?

Reply:
I think #587 can solve this problem.

Reply:
I don't have a good idea right now whether this situation should be included in this rule. Is there anyone who would skip calcite's parser and use the rules directly.